handshake_client.go

```diff
@@ -21,13 +21,14 @@ import (
 )
 
 type clientHandshakeState struct {
-	c            *Conn
-	serverHello  *serverHelloMsg
-	hello        *clientHelloMsg
-	suite        *cipherSuite
-	finishedHash finishedHash
-	masterSecret []byte
-	session      *ClientSessionState
+	c                *Conn
+	serverHello      *serverHelloMsg
+	hello            *clientHelloMsg
+	suite            *cipherSuite
+	finishedHash     finishedHash
+	masterSecret     []byte
+	session          *ClientSessionState
+	delCredPublicKey crypto.PublicKey
```

Comment:
Maybe this should be just connection PublicKey, and it is set to either the delegated cred key or c.peerCertificates[0].PublicKey. Then you may be able to reduce the number of if statements below.